doc/man/sparksd.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.1.
.TH SPARKS "1" "May 2025" "Sparks Core version v21.1.1" "User Commands"
.SH NAME
Sparks \- manual page for Sparks Core version v21.1.1
.SH SYNOPSIS
.B sparksd
[\fI\,options\/\fR]                     \fI\,Start Sparks Core\/\fR
.SH DESCRIPTION
Sparks Core version v21.1.1
Copyright \(co 2018\-2025 The Sparks Core developers
Copyright \(co 2014\-2025 The Dash Core developers
Copyright \(co 2009\-2025 The Bitcoin Core developers
.PP
Please contribute if you find Sparks Core useful. Visit <https://sparkspay.io/>
for further information about the software.
The source code is available from <https://github.com/sparkspay/sparks>.
.PP
This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/licenses/MIT>
.SH OPTIONS
.HP
\-?
.IP
Print this help message and exit
.HP
\fB\-alertnotify=\fR<cmd>
.IP
Execute command when an alert is raised (%s in cmd is replaced by
message)
.HP
\fB\-assumevalid=\fR<hex>
.IP
If this block is in the chain assume that it and its ancestors are valid
and potentially skip their script verification (0 to verify all,
default:
000000072b5fe92e4b1ef1014777dfd379352d67037b5888de2c26b60ad03b28,
testnet:
000009a6d2be24c757b06d5fa9b392b97f1b07d3847b49fdfaf84fe2d4c1412c)
.HP
\fB\-blockfilterindex=\fR<type>
.IP
Maintain an index of compact filters by block (default: 0, values:
basic). If <type> is not supplied or if <type> = 1, indexes for
all known types are enabled.
.HP
\fB\-blocknotify=\fR<cmd>
.IP
Execute command when the best block changes (%s in cmd is replaced by
block hash)
.HP
\fB\-blockreconstructionextratxn=\fR<n>
.IP
Extra transactions to keep in memory for compact block reconstructions
(default: 100)
.HP
\fB\-blocksdir=\fR<dir>
.IP
Specify directory to hold blocks subdirectory for *.dat files (default:
<datadir>)
.HP
\fB\-blocksonly\fR
.IP
Whether to reject transactions from network peers. Automatic broadcast
and rebroadcast of any transactions from inbound peers is
disabled, unless the peer has the 'forcerelay' permission. RPC
transactions are not affected. (default: 0)
.HP
\fB\-chainlocknotify=\fR<cmd>
.IP
Execute command when the best chainlock changes (%s in cmd is replaced
by chainlocked block hash)
.HP
\fB\-coinstatsindex\fR
.IP
Maintain coinstats index used by the gettxoutsetinfo RPC (default: 0)
.HP
\fB\-conf=\fR<file>
.IP
Specify path to read\-only configuration file. Relative paths will be
prefixed by datadir location. (default: sparks.conf)
.HP
\fB\-daemon\fR
.IP
Run in the background as a daemon and accept commands (default: 0)
.HP
\fB\-daemonwait\fR
.IP
Wait for initialization to be finished before exiting. This implies
\fB\-daemon\fR (default: 0)
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.HP
\fB\-dbcache=\fR<n>
.IP
Maximum database cache size <n> MiB (4 to 16384, default: 300). In
addition, unused mempool memory is shared for this cache (see
\fB\-maxmempool\fR).
.HP
\fB\-debuglogfile=\fR<file>
.IP
Specify location of debug log file. Relative paths will be prefixed by a
net\-specific datadir location. (\fB\-nodebuglogfile\fR to disable;
default: debug.log)
.HP
\fB\-includeconf=\fR<file>
.IP
Specify additional configuration file, relative to the \fB\-datadir\fR path
(only useable from configuration file, not command line)
.HP
\fB\-loadblock=\fR<file>
.IP
Imports blocks from external file on startup
.HP
\fB\-maxmempool=\fR<n>
.IP
Keep the transaction memory pool below <n> megabytes (default: 300)
.HP
\fB\-maxorphantxsize=\fR<n>
.IP
Maximum total size of all orphan transactions in megabytes (default: 10)
.HP
\fB\-maxrecsigsage=\fR<n>
.IP
Number of seconds to keep LLMQ recovery sigs (default: 604800)
.HP
\fB\-mempoolexpiry=\fR<n>
.IP
Do not keep transactions in the mempool longer than <n> hours (default:
336)
.HP
\fB\-par=\fR<n>
.IP
Set the number of script verification threads (\fB\-12\fR to 15, 0 = auto, <0 =
leave that many cores free, default: 0)
.HP
\fB\-persistmempool\fR
.IP
Whether to save the mempool on shutdown and load on restart (default: 1)
.HP
\fB\-pid=\fR<file>
.IP
Specify pid file. Relative paths will be prefixed by a net\-specific
datadir location. (default: sparksd.pid)
.HP
\fB\-prune=\fR<n>
.IP
Reduce storage requirements by enabling pruning (deleting) of old
blocks. This allows the pruneblockchain RPC to be called to
delete specific blocks, and enables automatic pruning of old
blocks if a target size in MiB is provided. This mode is
incompatible with \fB\-txindex\fR, \fB\-coinstatsindex\fR, \fB\-rescan\fR and
\fB\-disablegovernance\fR=\fI\,false\/\fR. Warning: Reverting this setting
requires re\-downloading the entire blockchain. (default: 0 =
disable pruning blocks, 1 = allow manual pruning via RPC, >945 =
automatically prune block files to stay under the specified
target size in MiB)
.HP
\fB\-settings=\fR<file>
.IP
Specify path to dynamic settings data file. Can be disabled with
\fB\-nosettings\fR. File is written at runtime and not meant to be
edited by users (use sparks.conf instead for custom settings).
Relative paths will be prefixed by datadir location. (default:
settings.json)
.HP
\fB\-startupnotify=\fR<cmd>
.IP
Execute command on startup.
.HP
\fB\-syncmempool\fR
.IP
Sync mempool from other nodes on start (default: 1)
.HP
\fB\-sysperms\fR
.IP
Create new files with system default permissions, instead of umask 077
(only effective with disabled wallet functionality)
.HP
\fB\-version\fR
.IP
Print version and exit
.PP
Connection options:
.HP
\fB\-addnode=\fR<ip>
.IP
Add a node to connect to and attempt to keep the connection open (see
the addnode RPC help for more info). This option can be specified
multiple times to add multiple nodes; connections are limited to
8 at a time and are counted separately from the \fB\-maxconnections\fR
limit.
.HP
\fB\-allowprivatenet\fR
.IP
Allow RFC1918 addresses to be relayed and connected to (default: 0)
.HP
\fB\-asmap=\fR<file>
.IP
Specify asn mapping used for bucketing of the peers (default:
ip_asn.map). Relative paths will be prefixed by the net\-specific
datadir location.
.HP
\fB\-bantime=\fR<n>
.IP
Default duration (in seconds) of manually configured bans (default:
86400)
.HP
\fB\-bind=\fR<addr>[:<port>][=onion]
.IP
Bind to given address and always listen on it (default: 0.0.0.0). Use
[host]:port notation for IPv6. Append =onion to tag any incoming
connections to that address and port as incoming Tor connections
(default: 127.0.0.1:9996=onion, testnet: 127.0.0.1:19996=onion,
regtest: 127.0.0.1:19896=onion)
.HP
\fB\-cjdnsreachable\fR
.IP
If set, then this host is configured for CJDNS (connecting to fc00::/8
addresses would lead us to the CJDNS network, see doc/cjdns.md)
(default: 0)
.HP
\fB\-connect=\fR<ip>
.IP
Connect only to the specified node; \fB\-noconnect\fR disables automatic
connections (the rules for this peer are the same as for
\fB\-addnode\fR). This option can be specified multiple times to connect
to multiple nodes.
.HP
\fB\-discover\fR
.IP
Discover own IP addresses (default: 1 when listening and no \fB\-externalip\fR
or \fB\-proxy\fR)
.HP
\fB\-dns\fR
.IP
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR (default: 1)
.HP
\fB\-dnsseed\fR
.IP
Query for peer addresses via DNS lookup, if low on addresses (default: 1
unless \fB\-connect\fR used)
.HP
\fB\-externalip=\fR<ip>
.IP
Specify your own public address
.HP
\fB\-fixedseeds\fR
.IP
Allow fixed seeds if DNS seeds don't provide peers (default: 1)
.HP
\fB\-forcednsseed\fR
.IP
Always query for peer addresses via DNS lookup (default: 0)
.HP
\fB\-i2pacceptincoming\fR
.IP
Whether to accept inbound I2P connections (default: 1). Ignored if
\fB\-i2psam\fR is not set. Listening for inbound I2P connections is done
through the SAM proxy, not by binding to a local address and
port.
.HP
\fB\-i2psam=\fR<ip:port>
.IP
I2P SAM proxy to reach I2P peers and accept I2P connections (default:
none)
.HP
\fB\-listen\fR
.IP
Accept connections from outside (default: 1 if no \fB\-proxy\fR or \fB\-connect\fR)
.HP
\fB\-listenonion\fR
.IP
Automatically create Tor onion service (default: 1)
.HP
\fB\-maxconnections=\fR<n>
.IP
Maintain at most <n> connections to peers (temporary service connections
excluded) (default: 125). This limit does not apply to
connections manually added via \fB\-addnode\fR or the addnode RPC, which
have a separate limit of 8.
.HP
\fB\-maxreceivebuffer=\fR<n>
.IP
Maximum per\-connection receive buffer, <n>*1000 bytes (default: 5000)
.HP
\fB\-maxsendbuffer=\fR<n>
.IP
Maximum per\-connection send buffer, <n>*1000 bytes (default: 1000)
.HP
\fB\-maxtimeadjustment\fR
.IP
Maximum allowed median peer time offset adjustment. Local perspective of
time may be influenced by peers forward or backward by this
amount. (default: 4200 seconds)
.HP
\fB\-maxuploadtarget=\fR<n>
.IP
Tries to keep outbound traffic under the given target (in MiB per 24h).
Limit does not apply to peers with 'download' permission. 0 = no
limit (default: 0)
.HP
\fB\-natpmp\fR
.IP
Use NAT\-PMP to map the listening port (default: 0)
.HP
\fB\-networkactive\fR
.IP
Enable all P2P network activity (default: 1). Can be changed by the
setnetworkactive RPC command
.HP
\fB\-onion=\fR<ip:port>
.IP
Use separate SOCKS5 proxy to reach peers via Tor onion services, set
\fB\-noonion\fR to disable (default: \fB\-proxy\fR)
.HP
\fB\-onlynet=\fR<net>
.IP
Make automatic outbound connections only to network <net> (ipv4, ipv6,
onion, i2p, cjdns). Inbound and manual connections are not
affected by this option. It can be specified multiple times to
allow multiple networks.
.HP
\fB\-peerblockfilters\fR
.IP
Serve compact block filters to peers per BIP 157 (default: 0)
.HP
\fB\-peerbloomfilters\fR
.IP
Support filtering of blocks and transaction with bloom filters (default:
1)
.HP
\fB\-peertimeout=\fR<n>
.IP
Specify a p2p connection timeout delay in seconds. After connecting to a
peer, wait this amount of time before considering disconnection
based on inactivity (minimum: 1, default: 60)
.HP
\fB\-permitbaremultisig\fR
.IP
Relay non\-P2SH multisig (default: 1)
.HP
\fB\-port=\fR<port>
.IP
Listen for connections on <port>. Nodes not using the default ports
(default: 8890, testnet: 8891, regtest: 18891) are unlikely to
get incoming connections. Not relevant for I2P (see doc/i2p.md).
.HP
\fB\-proxy=\fR<ip:port>
.IP
Connect through SOCKS5 proxy, set \fB\-noproxy\fR to disable (default:
disabled)
.HP
\fB\-proxyrandomize\fR
.IP
Randomize credentials for every proxy connection. This enables Tor
stream isolation (default: 1)
.HP
\fB\-seednode=\fR<ip>
.IP
Connect to a node to retrieve peer addresses, and disconnect. This
option can be specified multiple times to connect to multiple
nodes.
.HP
\fB\-socketevents=\fR<mode>
.IP
Socket events mode, which must be one of 'select', 'poll', 'epoll' or
\&'kqueue', depending on your system (default: Linux \- 'epoll',
FreeBSD/Apple \- 'kqueue', Windows \- 'select')
.HP
\fB\-timeout=\fR<n>
.IP
Specify socket connection timeout in milliseconds. If an initial attempt
to connect is unsuccessful after this amount of time, drop it
(minimum: 1, default: 5000)
.HP
\fB\-torcontrol=\fR<ip>:<port>
.IP
Tor control port to use if onion listening enabled (default:
127.0.0.1:9051)
.HP
\fB\-torpassword=\fR<pass>
.IP
Tor control port password (default: empty)
.HP
\fB\-upnp\fR
.IP
Use UPnP to map the listening port (default: 1 when listening and no
\fB\-proxy\fR)
.HP
\fB\-whitebind=\fR<[permissions@]addr>
.IP
Bind to the given address and add permission flags to the peers
connecting to it. Use [host]:port notation for IPv6. Allowed
permissions: bloomfilter (allow requesting BIP37 filtered blocks
and transactions), noban (do not ban for misbehavior; implies
download), forcerelay (relay transactions that are already in the
mempool; implies relay), relay (relay even in \fB\-blocksonly\fR mode),
mempool (allow requesting BIP35 mempool contents), download
(allow getheaders during IBD, no disconnect after maxuploadtarget
limit), addr (responses to GETADDR avoid hitting the cache and
contain random records with the most up\-to\-date info). Specify
multiple permissions separated by commas (default:
download,noban,mempool,relay). Can be specified multiple times.
.HP
\fB\-whitelist=\fR<[permissions@]IP address or network>
.IP
Add permission flags to the peers connecting from the given IP address
(e.g. 1.2.3.4) or CIDR\-notated network (e.g. 1.2.3.0/24). Uses
the same permissions as \fB\-whitebind\fR. Can be specified multiple
times.
.PP
Indexing options:
.HP
\fB\-addressindex\fR
.IP
Maintain a full address index, used to query for the balance, txids and
unspent outputs for addresses (default: 0)
.HP
\fB\-reindex\fR
.IP
Rebuild chain state and block index from the blk*.dat files on disk
.HP
\fB\-reindex\-chainstate\fR
.IP
Rebuild chain state from the currently indexed blocks. When in pruning
mode or if blocks on disk might be corrupted, use full \fB\-reindex\fR
instead.
.HP
\fB\-spentindex\fR
.IP
Maintain a full spent index, used to query the spending txid and input
index for an outpoint (default: 0)
.HP
\fB\-timestampindex\fR
.IP
Maintain a timestamp index for block hashes, used to query blocks hashes
by a range of timestamps (default: 0)
.HP
\fB\-txindex\fR
.IP
Maintain a full transaction index, used by the getrawtransaction rpc
call (default: 1)
.PP
Masternode options:
.HP
\fB\-llmq\-data\-recovery=\fR<n>
.IP
Enable automated quorum data recovery (default: 1)
.HP
\fB\-llmq\-qvvec\-sync=\fR<quorum_name>:<mode>
.IP
Defines from which LLMQ type the masternode should sync quorum
verification vectors. Can be used multiple times with different
LLMQ types. <mode>: 0 (sync always from all quorums of the type
defined by <quorum_name>), 1 (sync from all quorums of the type
defined by <quorum_name> if a member of any of the quorums)
.HP
\fB\-masternodeblsprivkey=\fR<hex>
.IP
Set the masternode BLS private key and enable the client to act as a
masternode
.HP
\fB\-platform\-user=\fR<user>
.IP
Set the username for the "platform user", a restricted user intended to
be used by Sparks Platform, to the specified username.
.PP
Statsd options:
.HP
\fB\-statsenabled\fR
.IP
Publish internal stats to statsd (default: 0)
.HP
\fB\-statshost=\fR<ip>
.IP
Specify statsd host (default: 127.0.0.1)
.HP
\fB\-statshostname=\fR<ip>
.IP
Specify statsd host name (default: )
.HP
\fB\-statsns=\fR<ns>
.IP
Specify additional namespace prefix (default: )
.HP
\fB\-statsperiod=\fR<seconds>
.IP
Specify the number of seconds between periodic measurements (default:
60)
.HP
\fB\-statsport=\fR<port>
.IP
Specify statsd port (default: 8125)
.PP
Wallet options:
.HP
\fB\-avoidpartialspends\fR
.IP
Group outputs by address, selecting many (possibly all) or none, instead
of selecting on a per\-output basis. Privacy is improved as
addresses are mostly swept with fewer transactions and outputs
are aggregated in clean change addresses. It may result in higher
fees due to less optimal coin selection caused by this added
limitation and possibly a larger\-than\-necessary number of inputs
being used. Always enabled for wallets with "avoid_reuse"
enabled, otherwise default: 0.
.HP
\fB\-createwalletbackups=\fR<n>
.IP
Number of automatic wallet backups (default: 10)
.HP
\fB\-disablewallet\fR
.IP
Do not load the wallet and disable wallet RPC calls
.HP
\fB\-instantsendnotify=\fR<cmd>
.IP
Execute command when a wallet InstantSend transaction is successfully
locked. %s in cmd is replaced by TxID and %w is replaced by
wallet name. %w is not currently implemented on Windows. On
systems where %w is supported, it should NOT be quoted because
this would break shell escaping used to invoke the command.
.HP
\fB\-keypool=\fR<n>
.IP
Set key pool size to <n> (default: 1000). Warning: Smaller sizes may
increase the risk of losing funds when restoring from an old
backup, if none of the addresses in the original keypool have
been used.
.HP
\fB\-maxapsfee=\fR<n>
.IP
Spend up to this amount in additional (absolute) fees (in Sparks) if it
allows the use of partial spend avoidance (default: 0.00)
.HP
\fB\-rescan=\fR<mode>
.IP
Rescan the block chain for missing wallet transactions on startup (1 =
start from wallet creation time, 2 = start from genesis block)
.HP
\fB\-spendzeroconfchange\fR
.IP
Spend unconfirmed change when sending transactions (default: 1)
.HP
\fB\-wallet=\fR<path>
.IP
Specify wallet path to load at startup. Can be used multiple times to
load multiple wallets. Path is to a directory containing wallet
data and log files. If the path is not absolute, it is
interpreted relative to <walletdir>. This only loads existing
wallets and does not create new ones. For backwards compatibility
this also accepts names of existing top\-level data files in
<walletdir>.
.HP
\fB\-walletbackupsdir=\fR<dir>
.IP
Specify full path to directory for automatic wallet backups (must exist)
.HP
\fB\-walletbroadcast\fR
.IP
Make the wallet broadcast transactions (default: 1)
.HP
\fB\-walletdir=\fR<dir>
.IP
Specify directory to hold wallets (default: <datadir>/wallets if it
exists, otherwise <datadir>)
.HP
\fB\-walletnotify=\fR<cmd>
.IP
Execute command when a wallet transaction changes. %s in cmd is replaced
by TxID and %w is replaced by wallet name. %w is not currently
implemented on windows. On systems where %w is supported, it
should NOT be quoted because this would break shell escaping used
to invoke the command.
.PP
Wallet fee options:
.HP
\fB\-discardfee=\fR<amt>
.IP
The fee rate (in Sparks/kB) that indicates your tolerance for discarding
change by adding it to the fee (default: 0.0001). Note: An output
is discarded if it is dust at this rate, but we will always
discard up to the dust relay fee and a discard fee above that is
limited by the fee estimate for the longest target
.HP
\fB\-fallbackfee=\fR<amt>
.IP
A fee rate (in Sparks/kB) that will be used when fee estimation has
insufficient data. 0 to entirely disable the fallbackfee feature.
(default: 0.00001)
.HP
\fB\-mintxfee=\fR<amt>
.IP
Fee rates (in Sparks/kB) smaller than this are considered zero fee for
transaction creation (default: 0.00001)
.HP
\fB\-paytxfee=\fR<amt>
.IP
Fee rate (in Sparks/kB) to add to transactions you send (default: 0.00)
.HP
\fB\-txconfirmtarget=\fR<n>
.IP
If paytxfee is not set, include enough fee so transactions begin
confirmation on average within n blocks (default: 6)
.PP
HD wallet options:
.HP
\fB\-hdseed=\fR<hex>
.IP
User defined seed for HD wallet (should be in hex). Only has effect
during wallet creation/first start (default: randomly generated)
.HP
\fB\-mnemonic=\fR<text>
.IP
User defined mnemonic for HD wallet (bip39). Only has effect during
wallet creation/first start (default: randomly generated)
.HP
\fB\-mnemonicbits=\fR<n>
.IP
User defined mnemonic security for HD wallet in bits (BIP39). Only has
effect during wallet creation/first start (allowed values: 128,
160, 192, 224, 256; default: 128)
.HP
\fB\-mnemonicpassphrase=\fR<text>
.IP
User defined mnemonic passphrase for HD wallet (BIP39). Only has effect
during wallet creation/first start (default: empty string)
.HP
\fB\-usehd\fR
.IP
Use hierarchical deterministic key generation (HD) after BIP39/BIP44.
Only has effect during wallet creation/first start (default: 1)
.PP
CoinJoin options:
.HP
\fB\-coinjoinamount=\fR<n>
.IP
Target CoinJoin balance (2\-21000000, default: 1000)
.HP
\fB\-coinjoinautostart\fR
.IP
Start CoinJoin automatically (0\-1, default: 0)
.HP
\fB\-coinjoindenomsgoal=\fR<n>
.IP
Try to create at least N inputs of each denominated amount (10\-100000,
default: 50)
.HP
\fB\-coinjoindenomshardcap=\fR<n>

doc/release-notes.md
# Sparks Core version v21.1.1

This is a new patch version release, bringing important bugfixes.

This release is **optional** but recommended for all nodes.

Please report bugs using the issue tracker at GitHub:

  <https://github.com/sparkspay/sparks/issues>


# Upgrading and downgrading

## How to Upgrade

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes for older versions), then run the
installer (on Windows) or just copy over /Applications/Sparks-Qt (on Mac) or
sparksd/sparks-qt (on Linux).

## Downgrade warning

### Downgrade to a version < v21.0.0

Downgrading to a version older than v21.0.0 may not be supported due to changes
if you are using descriptor wallets.

### Downgrade to a version < v19.2.0

Downgrading to a version older than v19.2.0 is not supported due to changes
in the evodb database. If you need to use an older version, you must either
reindex or re-sync the whole chain.

# Notable changes

- Core now categorizes asset unlock transactions as "Platform Transfers" on the Transactions tab in Sparks-Qt and in the output of the `gettransaction` RPC (#6131)
- Persist Coinjoin Denoms options changes made via GUI over restarts (#6208)
- Fix incorrect payment predictions for evonodes in Sparks-Qt and in RPC `masternode winners` (#6222)
- `creditOutputs` entries in various RPCs that output transaction JSON are shown as objects now instead of being shown as strings (#6229)
- Updated PGP key for builder 'pasta' to reflect new subkeys. You may need to reimport this key to validate signatures. (#6290)
- Build failures on Ubuntu 24.10 / clang 19.1.1 resolved (#6328)
- RPC errors in `masternode payments`, `getblock`, `getblockstats` related to Asset Unlock parsing have been fixed (#6336)

# v21.1.0 Change log

See detailed [set of changes][set-of-changes].

# Credits

Thanks to everyone who directly contributed to this release:

- Kittywhiskers Van Gogh
- Konstantin Akimov
- PastaPastaPasta
- UdjinM6

As well as everyone that submitted issues, reviewed pull requests and helped
debug the release candidates.

# Older releases

These release are considered obsolete. Old release notes can be found here:

- [v21.1.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-21.1.0.md) released Aug/8/2024
- [v21.0.2](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-21.0.2.md) released Aug/1/2024
- [v21.0.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-21.0.0.md) released Jul/25/2024
- [v20.1.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-20.1.1.md) released April/3/2024
- [v20.1.0](https://github.com/Sparkspay/Sparks/blob/master/doc/release-notes/Sparks/release-notes-20.1.0.md) released March/5/2024
- [v20.0.4](https://github.com/Sparkspay/Sparks/blob/master/doc/release-notes/Sparks/release-notes-20.0.4.md) released Jan/13/2024
- [v20.0.3](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-20.0.3.md) released December/26/2023
- [v20.0.2](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-20.0.2.md) released December/06/2023
- [v20.0.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-20.0.1.md) released November/18/2023
- [v20.0.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-20.0.0.md) released November/15/2023
- [v19.3.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-19.3.0.md) released July/31/2023
- [v18.2.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-18.2.2.md) released Jan/17/2023
- [v18.2.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-18.2.0.md) released Jan/01/2023
- [v18.1.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-18.1.1.md) released January/08/2023
- [v18.1.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-18.1.0.md) released October/09/2022
- [v18.0.2](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-18.0.2.md) released October/09/2022
- [v18.0.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-18.0.1.md) released August/17/2022
- [v0.17.0.3](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.17.0.3.md) released June/07/2021
- [v0.17.0.2](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.17.0.2.md) released May/19/2021
- [v0.16.1.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.16.1.1.md) released November/17/2020
- [v0.16.1.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.16.1.0.md) released November/14/2020
- [v0.16.0.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.16.0.1.md) released September/30/2020
- [v0.15.0.0](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.15.0.0.md) released Febrary/18/2020
- [v0.14.0.5](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.14.0.5.md) released December/08/2019
- [v0.14.0.4](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.14.0.4.md) released November/22/2019
- [v0.14.0.3](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.14.0.3.md) released August/15/2019
- [v0.14.0.2](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.14.0.2.md) released July/4/2019
- [v0.14.0.1](https://github.com/sparkspay/sparks/blob/master/doc/release-notes/sparks/release-notes-0.14.0.1.md) released May/31/2019

src/sparksd.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2015 The Bitcoin Core developers
// Copyright (c) 2014-2022 The Dash Core developers
// Copyright (c) 2016-2025 The Sparks Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#if defined(HAVE_CONFIG_H)
#endif

#include <chainparams.h>
#include <clientversion.h>
#include <compat.h>
#include <fs.h>
#include <rpc/server.h>
#include <init.h>
#include <noui.h>
#include <httpserver.h>
#include <httprpc.h>
#include <walletinitinterface.h>
#include <stacktraces.h>

#include <stdio.h>

/* Introduction text for doxygen: */

/*! \mainpage Developer documentation
 *
 * \section intro_sec Introduction
 *
 * This is the developer documentation of the reference client for an experimental new digital currency called Sparks (https://www.sparkspay.io/),
 * which enables instant payments to anyone, anywhere in the world. Sparks uses peer-to-peer technology to operate
 * with no central authority: managing transactions and issuing money are carried out collectively by the network.
 *
 * The software is a community-driven open source project, released under the MIT license.
 *
 * \section Navigation
 * Use the buttons <code>Namespaces</code>, <code>Classes</code> or <code>Files</code> at the top of the page to start navigating the code.
 */

void WaitForShutdown()
{
    while (!ShutdownRequested())
    {
        MilliSleep(200);
    }
    Interrupt();
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//
bool AppInit(int argc, char* argv[])
{
    bool fRet = false;

    //
    // Parameters
    //
    // If Qt is used, parameters/sparks.conf are parsed in qt/sparks.cpp's main()
    SetupServerArgs();
#if HAVE_DECL_DAEMON
    gArgs.AddArg("-daemon", "Run in the background as a daemon and accept commands", false, OptionsCategory::OPTIONS);
#endif
    gArgs.ParseParameters(argc, argv);

    if (gArgs.IsArgSet("-printcrashinfo")) {
        std::cout << GetCrashInfoStrFromSerializedStr(gArgs.GetArg("-printcrashinfo", "")) << std::endl;
        return true;
    }

    // Process help and version before taking care about datadir
    if (gArgs.IsArgSet("-?") || gArgs.IsArgSet("-h") ||  gArgs.IsArgSet("-help") || gArgs.IsArgSet("-version"))
    {
        std::string strUsage = strprintf("%s Daemon", PACKAGE_NAME) + " version " + FormatFullVersion() + "\n";

        if (gArgs.IsArgSet("-version"))
        {
            strUsage += FormatParagraph(LicenseInfo());
        }
        else
        {
            strUsage += "\nUsage:\n"
                  "  sparksd [options]                     " + strprintf("Start %s Daemon", PACKAGE_NAME) + "\n";

            strUsage += "\n" + gArgs.GetHelpMessage();
        }

        fprintf(stdout, "%s", strUsage.c_str());
        return true;
    }

    try
    {
        bool datadirFromCmdLine = gArgs.IsArgSet("-datadir");
        if (datadirFromCmdLine && !fs::is_directory(GetDataDir(false)))
        {
            fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
            return false;
        }
        try
        {
            gArgs.ReadConfigFile(gArgs.GetArg("-conf", BITCOIN_CONF_FILENAME));
        } catch (const std::exception& e) {
            fprintf(stderr,"Error reading configuration file: %s\n", e.what());
            return false;
        }
        if (!datadirFromCmdLine && !fs::is_directory(GetDataDir(false)))
        {
            fprintf(stderr, "Error: Specified data directory \"%s\" from config file does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
            return EXIT_FAILURE;
        }
        // Check for -testnet or -regtest parameter (Params() calls are only valid after this clause)
        try {
            SelectParams(gArgs.GetChainName());
        } catch (const std::exception& e) {
            fprintf(stderr, "Error: %s\n", e.what());
            return false;
        }

        // Error out when loose non-argument tokens are encountered on command line
        for (int i = 1; i < argc; i++) {
            if (!IsSwitchChar(argv[i][0])) {
                fprintf(stderr, "Error: Command line contains unexpected token '%s', see sparksd -h for a list of options.\n", argv[i]);
                return false;
            }
        }

        // -server defaults to true for sparksd but not for the GUI so do this here
        gArgs.SoftSetBoolArg("-server", true);
        // Set this early so that parameter interactions go to console
        InitLogging();
        InitParameterInteraction();
        if (!AppInitBasicSetup())
        {
            // InitError will have been called with detailed error, which ends up on console
            return false;
        }
        if (!AppInitParameterInteraction())
        {
            // InitError will have been called with detailed error, which ends up on console
            return false;
        }
        if (!AppInitSanityChecks())
        {
            // InitError will have been called with detailed error, which ends up on console
            return false;
        }
        if (gArgs.GetBoolArg("-daemon", false))
        {
#if HAVE_DECL_DAEMON
#if defined(MAC_OSX)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#endif
            fprintf(stdout, "Sparks Core server starting\n");

            // Daemonize
            if (daemon(1, 0)) { // don't chdir (1), do close FDs (0)
                fprintf(stderr, "Error: daemon() failed: %s\n", strerror(errno));
                return false;
            }
#if defined(MAC_OSX)
#pragma GCC diagnostic pop
#endif
#else
            fprintf(stderr, "Error: -daemon is not supported on this operating system\n");
            return false;
#endif // HAVE_DECL_DAEMON
        }
        // Lock data directory after daemonization
        if (!AppInitLockDataDirectory())
        {
            // If locking the data directory failed, exit immediately
            return false;
        }
        fRet = AppInitMain();
    } catch (...) {
        PrintExceptionContinue(std::current_exception(), "AppInit()");
    }

    if (!fRet)
    {
        Interrupt();
    } else {
        WaitForShutdown();
    }
    Shutdown();

    return fRet;
}

int main(int argc, char* argv[])
{
    RegisterPrettyTerminateHander();
    RegisterPrettySignalHandlers();

    SetupEnvironment();

    // Connect sparksd signal handlers
    noui_connect();

    return (AppInit(argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
}
